src/states/common/relocated.rs

```diff
@@ -0,0 +1,357 @@
+            .map_err(RoutingError::Crypto)
+    }
+
+    // If `msg_id` is `Some` this is sent as a response, otherwise as a request.
+    fn send_connection_info(
+        &mut self,
+        our_pub_info: PubConnectionInfo,
+        their_pub_id: PublicId,
+        src: Authority<XorName>,
+        dst: Authority<XorName>,
+        msg_id: Option<MessageId>,
+    ) {
+        let shared_secret = self
+            .full_id()
+            .encrypting_private_key()
+            .shared_secret(&their_pub_id.encrypting_public_key());
+        let encrypted_conn_info = match shared_secret.encrypt(&our_pub_info) {
+            Ok(encrypted_conn_info) => encrypted_conn_info,
+            Err(err) => {
+                debug!(
+                    "{} Failed to serialise connection info for {:?}: {:?}.",
+                    self, their_pub_id, err
+                );
+                return;
+            }
+        };
+        let msg_content = if let Some(msg_id) = msg_id {
+            MessageContent::ConnectionInfoResponse {
+                encrypted_conn_info,
+                pub_id: *self.full_id().public_id(),
+                msg_id,
+            }
+        } else {
+            MessageContent::ConnectionInfoRequest {
+                encrypted_conn_info,
+                pub_id: *self.full_id().public_id(),
+                msg_id: MessageId::new(),
+            }
+        };
+
+        if let Err(err) = self.send_routing_message(src, dst, msg_content) {
+            debug!(
+                "{} Failed to send connection info for {:?}: {:?}.",
+                self, their_pub_id, err
+            );
+        }
+    }
+
+    /// Disconnects if the peer is not a proxy, client or routing table entry.
+    fn disconnect_peer(&mut self, pub_id: &PublicId) {
+        if self
+            .peer_mgr()
+            .get_peer(pub_id)
+            .map_or(false, Peer::is_routing)
+        {
+            debug!("{} Not disconnecting routing table entry {}.", self, pub_id);
+        } else if self.peer_mgr().is_proxy(pub_id) {
+            debug!("{} Not disconnecting proxy node {}.", self, pub_id);
+        } else if self.peer_mgr().is_joining_node(pub_id) {
+            debug!("{} Not disconnecting joining node {:?}.", self, pub_id);
+        } else {
+            debug!(
+                "{} Disconnecting {}. Calling crust::Service::disconnect.",
+                self, pub_id
+            );
+            let _ = self.crust_service().disconnect(pub_id);
+            let _ = self.peer_mgr().remove_peer(pub_id);
+        }
+    }
+
+    // Filter, then convert the message to a `Hop` and serialise.
+    // Send this byte string.
+    fn send_signed_message_to_peer(
+        &mut self,
+        signed_msg: SignedMessage,
+        target: &PublicId,
+        route: u8,
+        sent_to: BTreeSet<XorName>,
+    ) -> Result<(), RoutingError> {
+        if self.filter_outgoing_routing_msg(signed_msg.routing_message(), target, route) {
```

Comment:
Why move the filter? I mean it make sense to me to do filter first, but it also seem to change things like before, it would have called disconnect_peer but it won't now.

Reply:
I think I wanted to avoid having to `clone` the routing_msg, but you are right, it changes the logic. Will fix.